File: bin/carve/render.py
#!/usr/bin/env python3
"""sprite.json の文字格子を PNG に描く (標準ライブラリのみ)。

  python3 bin/render.py                 # assets/ 配下を全部 gallery/ へ
  python3 bin/render.py a.sprite.json   # 指定ファイルだけ

色は Doc 直下の inline "palette" (名前 → "#rrggbb") で解く。
拡大率は絵の大きさで自動 (48px 以下 = 8 倍、それ以上 = 4 倍)。
"""

import json
import os
import struct
import sys
import zlib

# Windows のコンソール (cp932 等) でも日本語メッセージで落とさない。
for stream in (sys.stdout, sys.stderr):
    if hasattr(stream, "reconfigure"):
        stream.reconfigure(encoding="utf-8", errors="replace")

TRANSPARENT_CHARS = {".", " "}


def png_of(width, height, rows_rgba):
    def chunk(tag, data):
        body = tag + data
        return struct.pack(">I", len(data)) + body + struct.pack(">I", zlib.crc32(body))

    raw = b"".join(
        b"\x00" + b"".join(struct.pack("4B", *px) for px in row) for row in rows_rgba
    )
    return (
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0))
        + chunk(b"IDAT", zlib.compress(raw, 9))
        + chunk(b"IEND", b"")
    )


def rgba_of(value):
    body = value[1:] if value.startswith("#") else value
    return tuple(int(body[i : i + 2], 16) for i in (0, 2, 4)) + (255,)


def hex_of(value):
    if isinstance(value, str):
        body = value[1:] if value.startswith("#") else value
        if len(body) == 6 and all(c in "0123456789abcdefABCDEF" for c in body):
            return "#" + body.lower()
    if isinstance(value, dict):
        for key in ("hex", "color", "value"):
            if key in value:
                found = hex_of(value[key])
                if found:
                    return found
    return None


def render_doc(path, out_dir):
    with open(path, encoding="utf-8") as handle:
        doc = json.load(handle)
    palette = {
        name: hex_of(value)
        for name, value in (doc.get("palette") or {}).items()
        if hex_of(value)
    }
    legend = {}
    for char, value in (doc.get("legend") or {}).items():
        direct = hex_of(value)
        name = value[1:] if isinstance(value, str) and value.startswith("@") else value
        resolved = direct or palette.get(name)
        if resolved:
            legend[char] = rgba_of(resolved)

    base = os.path.basename(path).replace(".sprite.json", "")
    written = []
    for sprite_name, spec in (doc.get("sprites") or {}).items():
        if sprite_name.startswith("//") or not isinstance(spec, dict):
            continue
        for frame_name, rows in (spec.get("frames") or {}).items():
            width = max((len(r) for r in rows), default=0)
            height = len(rows)
            if not width or not height:
                continue
            scale = 8 if max(width, height) <= 48 else 4
            rgba_rows = []
            for row in rows:
                line = []
                for x in range(width):
                    char = row[x] if x < len(row) else "."
                    line.extend([legend.get(char, (0, 0, 0, 0))] * scale)
                rgba_rows.extend([line] * scale)
            out = os.path.join(out_dir, f"{base}__{sprite_name}__{frame_name}.png")
            with open(out, "wb") as handle:
                handle.write(png_of(width * scale, height * scale, rgba_rows))
            written.append(out)
    return written


def main(argv):
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    out_dir = os.path.join(root, "gallery")
    os.makedirs(out_dir, exist_ok=True)
    targets = argv or sorted(
        os.path.join(current, name)
        for current, _, files in os.walk(os.path.join(root, "assets"))
        for name in files
        if name.endswith(".sprite.json")
    )
    count = 0
    for path in targets:
        for out in render_doc(path, out_dir):
            print(os.path.relpath(out, root))
            count += 1
    print(f"{count} 枚を書き出した")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: go/internal/carve/render.go
package carve

// bin/carve/render.py の写し。sprite.json の文字格子を PNG に描く。
//
//	fge-go carve-render                 # bin/assets 配下を全部 bin/gallery へ
//	fge-go carve-render a.sprite.json   # 指定ファイルだけ

import (
	"bytes"
	"compress/zlib"
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// RGBA は 1 画素。
type RGBA [4]int

// RGB は不透明な色。
type RGB [3]int

func pngChunk(buf *bytes.Buffer, tag string, data []byte) {
	_ = binary.Write(buf, binary.BigEndian, uint32(len(data)))
	body := append([]byte(tag), data...)
	buf.Write(body)
	_ = binary.Write(buf, binary.BigEndian, crc32.ChecksumIEEE(body))
}

// PNGOf は RGBA の格子を PNG のバイト列にする。
//
// WhyNot: 出たバイト列は Python 版と同一にならない (zlib と compress/flate は
// 同じ画素から違う圧縮結果を作る)。合わせられるのは描かれた画素まで。
func PNGOf(width, height int, rows [][]RGBA) []byte {
	raw := make([]byte, 0, height*(1+width*4))
	for _, row := range rows {
		raw = append(raw, 0)
		for _, px := range row {
			raw = append(raw, byte(px[0]), byte(px[1]), byte(px[2]), byte(px[3]))
		}
	}
	var z bytes.Buffer
	w, _ := zlib.NewWriterLevel(&z, zlib.BestCompression)
	_, _ = w.Write(raw)
	_ = w.Close()

	var buf bytes.Buffer
	buf.Write([]byte{0x89, 'P', 'N', 'G', '\r', '\n', 0x1a, '\n'})
	ihdr := make([]byte, 13)
	binary.BigEndian.PutUint32(ihdr[0:], uint32(width))
	binary.BigEndian.PutUint32(ihdr[4:], uint32(height))
	ihdr[8], ihdr[9] = 8, 6
	pngChunk(&buf, "IHDR", ihdr)
	pngChunk(&buf, "IDAT", z.Bytes())
	pngChunk(&buf, "IEND", nil)
	return buf.Bytes()
}

// RGBAOf は "#rrggbb" を不透明な画素にする。
func RGBAOf(value string) RGBA {
	body := value
	if strings.HasPrefix(body, "#") {
		body = body[1:]
	}
	out := RGBA{0, 0, 0, 255}
	for i, at := range []int{0, 2, 4} {
		v, _ := strconv.ParseInt(body[at:at+2], 16, 32)
		out[i] = int(v)
	}
	return out
}

// HexOf は値から "#rrggbb" を取り出す。取れなければ空。
func HexOf(value any) string {
	if s, ok := value.(string); ok {
		body := s
		if strings.HasPrefix(body, "#") {
			body = body[1:]
		}
		if len(body) == 6 && strings.IndexFunc(body, func(r rune) bool {
			return !strings.ContainsRune("0123456789abcdefABCDEF", r)
		}) < 0 {
			return "#" + strings.ToLower(body)
		}
	}
	if m, ok := value.(*OMap[string, any]); ok {
		for _, key := range []string{"hex", "color", "value"} {
			if v, ok := m.Get(key); ok {
				if found := HexOf(v); found != "" {
					return found
				}
			}
		}
	}
	return ""
}

// RenderDoc は 1 つの sprite.json を PNG に描き、書いたパスを返す。
func RenderDoc(path, outDir string) ([]string, error) {
	doc, err := readJSONOrdered(path)
	if err != nil {
		return nil, err
	}
	palette := map[string]string{}
	if p, ok := doc.Get("palette"); ok {
		if obj, ok := p.(*OMap[string, any]); ok {
			for _, name := range obj.Keys() {
				v, _ := obj.Get(name)
				if hex := HexOf(v); hex != "" {
					palette[name] = hex
				}
			}
		}
	}
	legend := map[string]RGBA{}
	if p, ok := doc.Get("legend"); ok {
		if obj, ok := p.(*OMap[string, any]); ok {
			for _, char := range obj.Keys() {
				v, _ := obj.Get(char)
				direct := HexOf(v)
				name := ""
				if s, ok := v.(string); ok {
					name = s
					if strings.HasPrefix(s, "@") {
						name = s[1:]
					}
				}
				resolved := direct
				if resolved == "" {
					resolved = palette[name]
				}
				if resolved != "" {
					legend[char] = RGBAOf(resolved)
				}
			}
		}
	}

	base := strings.Replace(filepath.Base(path), ".sprite.json", "", 1)
	var written []string
	sprites, _ := doc.Get("sprites")
	spriteObj, ok := sprites.(*OMap[string, any])
	if !ok {
		return written, nil
	}
	for _, spriteName := range spriteObj.Keys() {
		specAny, _ := spriteObj.Get(spriteName)
		spec, isObj := specAny.(*OMap[string, any])
		if strings.HasPrefix(spriteName, "//") || !isObj {
			continue
		}
		framesAny, _ := spec.Get("frames")
		frames, isObj := framesAny.(*OMap[string, any])
		if !isObj {
			continue
		}
		for _, frameName := range frames.Keys() {
			rowsAny, _ := frames.Get(frameName)
			rowsList, isList := rowsAny.([]any)
			if !isList {
				continue
			}
			var rows []string
			for _, r := range rowsList {
				s, _ := r.(string)
				rows = append(rows, s)
			}
			width := 0
			for _, r := range rows {
				if n := len([]rune(r)); n > width {
					width = n
				}
			}
			height := len(rows)
			if width == 0 || height == 0 {
				continue
			}
			side := width
			if height > side {
				side = height
			}
			scale := 4
			if side <= 48 {
				scale = 8
			}
			var rgbaRows [][]RGBA
			for _, row := range rows {
				runes := []rune(row)
				var line []RGBA
				for x := 0; x < width; x++ {
					char := "."
					if x < len(runes) {
						char = string(runes[x])
					}
					px := RGBA{0, 0, 0, 0}
					if c, ok := legend[char]; ok {
						px = c
					}
					for i := 0; i < scale; i++ {
						line = append(line, px)
					}
				}
				for i := 0; i < scale; i++ {
					rgbaRows = append(rgbaRows, line)
				}
			}
			out := filepath.Join(outDir, fmt.Sprintf("%s__%s__%s.png", base, spriteName, frameName))
			if err := os.WriteFile(out, PNGOf(width*scale, height*scale, rgbaRows), 0o644); err != nil {
				return written, err
			}
			written = append(written, out)
		}
	}
	return written, nil
}

// RunRender は render.py の入口。
func RunRender(out *strings.Builder, root string, args []string) (int, error) {
	outDir := filepath.Join(root, "gallery")
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		return 1, err
	}
	targets := args
	if len(targets) == 0 {
		var found []string
		_ = filepath.Walk(filepath.Join(root, "assets"), func(p string, info os.FileInfo, err error) error {
			if err != nil || info.IsDir() {
				return nil //nolint:nilerr // 走査できない枝は Python の os.walk と同じく黙って飛ばす
			}
			if strings.HasSuffix(p, ".sprite.json") {
				found = append(found, p)
			}
			return nil
		})
		sort.Strings(found)
		targets = found
	}
	count := 0
	for _, path := range targets {
		written, err := RenderDoc(path, outDir)
		if err != nil {
			return 1, err
		}
		for _, w := range written {
			rel, _ := filepath.Rel(root, w)
			fmt.Fprintln(out, rel)
			count++
		}
	}
	fmt.Fprintf(out, "%d 枚を書き出した\n", count)
	return 0, nil
}
